Hakawai/Mentions/HKWMentionsPlugin.m

```diff
@@ -824,6 +844,54 @@ - (void)toggleAutocorrectAsRequiredForRange:(NSRange)range {
     }
 }
 
+#pragma mark - Mention Query Utils
+
+- (NSString *)mentionsQuery:(NSString *)text location:(NSUInteger)location {
+    if (text.length <= 0) {
+        return nil;
+    }
+    // Search starting from the given location, and return first control character
+    NSUInteger mostRecentValidControlCharacterLocation = [self mostRecentValidControlCharacterLocation:text beforeLocation:location];
+    if (mostRecentValidControlCharacterLocation != NSNotFound) {
+        // Query until end of word in which cursor is present (or until cursor if it is at end of word)
+        NSString *const wordAfterCurrentLocation = [HKWMentionsStartDetectionStateMachine wordAfterLocation:location text:text];
+        NSString *substringUntilEndOfWord = [text substringToIndex:location+wordAfterCurrentLocation.length];
+        // Return the rest of the string after the control char as the query
+        NSString *query = [substringUntilEndOfWord substringFromIndex:mostRecentValidControlCharacterLocation + 1];
+        return query;
+    }
+    return nil;
+}
+
+- (NSUInteger)mostRecentValidControlCharacterLocation:(NSString *)text beforeLocation:(NSUInteger)location {
+    NSString *substringUntilLocation = [text substringToIndex:location];
+    // Search back MAX_MENTION_QUERY_LENGTH for a control character
+    NSUInteger maximumSearchIndex = (NSUInteger)MAX((int)location-MAX_MENTION_QUERY_LENGTH, 0);
+    NSString *substringToSearchForControlChar = [substringUntilLocation substringFromIndex:maximumSearchIndex];
+    // TODO: use self.controlCharacterSet and find most recent of any control char
+    // JIRA: POST-13613
+    NSRange rangeOfControlChar = [substringToSearchForControlChar rangeOfString:@"@" options:NSBackwardsSearch];
+    NSUInteger controlCharLocation = rangeOfControlChar.location;
+
+    // If there's a non-mentions alphanumeric before the control char, then it's invalid
+    unichar charPrecidingControlChar = [self.parentTextView characterPrecedingLocation:(NSInteger)controlCharLocation];
+    if (charPrecidingControlChar
+        && [[NSCharacterSet alphanumericCharacterSet] characterIsMember:charPrecidingControlChar]
+        && ![self mentionAttributeAtLocation:controlCharLocation-1
+                                       range:nil]) {
+        return NSNotFound;
+    }
+
+    // If there's an entity in between the current location and the previous control character, then it's invalid
+    NSRange mentionRange;
+    HKWMentionsAttribute __unused *mention = [self mentionAttributePrecedingLocation:location
+                                                                               range:&mentionRange];
+    if (mentionRange.location > controlCharLocation) {
+        return NSNotFound;
+    }
+
+    return controlCharLocation;
+}
```

Comment:
since they are utils, shall we make them class methods

Reply:
also would be great if we can add some comments to the function

Reply:
will turn to class methods in next rb. non-trivial and would affect other unrelated callsites currently ineffeciently using self in related class methods